internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// tlsstate manages TLS keypairs on behalf of the daemon's HTTPS server.
+package tlsstate
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha512"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base32"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/fs"
+	"math/big"
+	"os"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	// systemTime can be faked during testing.
+	systemTime = time.Now
+
+	// certificateValidityPeriod determines how long a newly generated
+	// TLS keypair is valid for.
+	certificateValidityPeriod = (10 * 365 * 24 * time.Hour)
+)
+
+type TLSManager struct {
+	// tlsDir is the location of the PEM keypair files.
+	tlsDir         string
+	tlsKeyPairLock sync.Mutex
+	tlsKeyPairs    []*TLSKeyPair
```

Comment:
The order that this slice is assumed to be in should be precisely documented here.
It's a very tight coupling in the design, so it should be hard to miss.

Reply:
I will add a comment.

Reply:
PR reworked. N/A for current PR. Resolving.